content/technotes/2020-05-28-rsvg2/index.md

```diff
@@ -0,0 +1,157 @@
+---
+slug: rsvg2
+title: 'Superior svg graphics rendering in R, and why it matters'
+date: '2020-05-28'
+author:
+  - Jeroen Ooms
+cover: https://docs.ropensci.org/av/reference/read_audio-1.png
+tags:
+  - R
+  - images
+  - packages
+  - svg
+---
+
+This week we released a major new version of the [rsvg](https://cran.r-project.org/web/packages/rsvg/index.html) package on CRAN. This package provides R bindings to [librsvg2](https://developer.gnome.org/rsvg/stable/) which is a powerful system library for rendering svg images into bitmaps, which we can use for further processing in for example the [magick](https://docs.ropensci.org/magick) package.
+
+The biggest change in this release is the R package on Windows and MacOS now includes the latest librsvg 2.48.4. This is a major upgrade; the librsvg2 rendering engine has been [completely rewritten](https://people.gnome.org/~federico/blog/css-in-librsvg-is-now-in-rust.html) in Rust using components from [Mozilla Servo](https://research.mozilla.org/servo-engines/). This has resulted in major improvements in quality and performance, and we have gained full support for css styling.
+
+In this post we showcase how it works, and why you should use svg for R graphics.
+
+## What is rendering
+
+A figure in svg format is stored as xml data containing a vector representation of a drawing, such as a sequence of lines, shapes, text, with their relative position, color, attributes, and so on. The benefit of storing graphics in svg format is that they can be resized without loss of quality. And because it is just xml, the shapes and text can be manipulated using standard xml/css tools, such as a browser.
+
+For an image to be displayed on screen, printed in a document, or loaded in editing software, it has to be _rendered_ into a bitmap. A bitmap is a fixed a array (matrix) of pixels with color values. Bitmap formats such as png, jpeg, or tiff all store the same pixel data, using different compression methods.
+
+The rsvg package renders svg into a bitmap image with the format and size of your choice, directly in R, and without loss of quality:
+
+```r
+# Example SVG image
+svgdata <- '<svg viewBox="0 0 1200 250" xmlns="http://www.w3.org/2000/svg">
+  <style>
+    circle {
+      fill: gold;
+      stroke: maroon;
+      stroke-width: 12px;
+    }
+    text {
+      fill: navy;
+      font-size: 2em;
+      font-family: "Times, Serif"
+    }
+  </style>	
+  <circle cx="200" cy="125" r="120" />
+  <text x="140" y="40" transform="rotate(30 10,20)">I love SVG!</text>
+</svg>'
+
+# Render with rsvg into png
+writeLines(svgdata, 'image.svg')
+rsvg::rsvg_png('image.svg', 'image.png', width = 800)
+```
+
+Instead of rendering to a png/jpeg file, you can also render the svg into raw bitmap data (called raw vectors in R), which you can read with for example [magick](https://docs.ropensci.org/magick) or any other imaging tool:
+
+
+```r
```

Comment:
maybe add a line on why there are 4 layers in the bitmap?

Reply:
Actually there can be any number of layers, depending on the colorspace. But for simplicity I just refer to each pixel having a color, without elaborating on how the color is made up.

Reply:
right but in the example shown with `str()` it'd be nice to explain the 4 "dimensions"